kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/Neo4jService.kt

```diff
@@ -112,33 +113,67 @@ class Neo4jService(private val config: Neo4jSinkConnectorConfig):
         }
     }
 
+    // taken from https://stackoverflow.com/questions/52192752/kotlin-how-to-run-n-coroutines-and-wait-for-first-m-results-or-timeout
+    @ObsoleteCoroutinesApi
+    @ExperimentalCoroutinesApi
+    suspend fun <T> List<Deferred<T>>.awaitAll(timeoutMs: Long): List<T> {
+        val jobs = CopyOnWriteArraySet<Deferred<T>>(this)
+        val result = ArrayList<T>(size)
+        val timeout = ticker(timeoutMs)
 
-    suspend fun writeData(data: Map<String, List<List<StreamsSinkEntity>>>) = coroutineScope {
-        val timeout = config.batchTimeout
-        val ticker = ticker(timeout)
-        val deferredList = data
-                .flatMap { (topic, records) ->
-                    records.map { async(Dispatchers.IO) { writeForTopic(topic, it) } }
-                }
         whileSelect {
-            ticker.onReceive {
-                if (log.isDebugEnabled) {
-                    log.debug("Timeout $timeout occurred while executing queries")
+            jobs.forEach { deferred ->
+                deferred.onAwait {
+                    jobs.remove(deferred)
+                    result.add(it)
+                    result.size != size
                 }
-                deferredList.forEach { deferred -> deferred.cancel() }
-                false // Stops the whileSelect
             }
-            val isAllCompleted = deferredList.all { it.isCompleted } // when all are completed
-            deferredList.forEach {
-                it.onAwait { !isAllCompleted } // Stops the whileSelect
+
+            timeout.onReceive {
+                jobs.forEach { it.cancel() }
+                throw TimeoutException("Tasks ${size} cancelled after timeout of $timeoutMs ms.")
             }
         }
-        val exceptionMessages = deferredList
-                .mapNotNull { it.getCompletionExceptionOrNull() }
-                .map { it.message }
-                .joinToString("\n")
-        if (exceptionMessages.isNotBlank()) {
-            throw ConnectException(exceptionMessages)
-        }
+
+        return result
     }
+
+    @ExperimentalCoroutinesApi
+    fun <T> Deferred<T>.errors() = when {
+        isCompleted -> getCompletionExceptionOrNull()
+        isCancelled -> getCompletionExceptionOrNull() // was getCancellationException()
+        isActive -> RuntimeException("Job $this still active")
+        else -> null }
+
+        @ExperimentalCoroutinesApi
+        @ObsoleteCoroutinesApi
+        suspend fun writeData(data: Map<String, List<List<StreamsSinkEntity>>>) = coroutineScope {
+
+            val errors = if (config.parallelBatches) {
+                val jobs = data
```

Comment:
Minor but it would be easier to read if you refactor this out to `writeDataAsync` which simplifies the logic of this writeData function.     Basically then it's just `val errors = config.parallelBatches ? writeDataAsync() : writeDataSync()`.   This also puts the writeDataAsync function on a peer level with the other and makes it easier to see the differences and unit test separately.

Reply:
Should we be concerned about the use of experimental APIs here? I defer to you because I don't know enough about Kotlin here.

Also, just as a Kotlin noob it'd be nice to have a line of comments in either method (sync/async) describing thread behavior and how/when this returns.   I.e. I think from reading this code that in the async method you're going to consume other threads (which is fine) but that the function overall doesn't return until they've all joined.

Reply:
it was already experimental APIs (coroutines) just added the annotations to remove warnings.

Reply:
it's a suspend method so it creates internal control structures to handle that correctly

basically in the middle of the method in several places it can give control back to the scheduler with the state of the method execution captured in a continuation.
it only `looks` like a "sync" method.

Reply:
Will do the other changes.